container-chains/nodes/frontier/src/rpc/finality.rs

```diff
@@ -0,0 +1,73 @@
+use fc_rpc::frontier_backend_client::{self, is_canon};
+use jsonrpsee::{core::RpcResult, proc_macros::rpc};
+use sp_blockchain::HeaderBackend;
+use sp_core::H256;
+use sp_runtime::traits::Block;
+use std::{marker::PhantomData, sync::Arc};
+
+#[rpc(server)]
+#[async_trait::async_trait]
+pub trait FrontierFinalityApi {
+    /// Reports whether a Substrate or Ethereum block is finalized.
+    /// Returns false if the block is not found.
+    #[method(name = "frnt_isBlockFinalized")]
+    fn is_block_finalized(&self, block_hash: H256) -> RpcResult<bool>;
+}
+
+pub struct FrontierFinality<B: Block, C> {
+    pub backend: Arc<dyn fc_api::Backend<B>>,
+    pub client: Arc<C>,
+    _phdata: PhantomData<B>,
+}
+
+impl<B: Block, C> FrontierFinality<B, C> {
+    pub fn new(client: Arc<C>, backend: Arc<dyn fc_api::Backend<B>>) -> Self {
+        Self {
+            backend,
+            client,
+            _phdata: Default::default(),
+        }
+    }
+}
+
+impl<B, C> FrontierFinalityApiServer for FrontierFinality<B, C>
+where
+    B: Block<Hash = H256>,
+    C: HeaderBackend<B> + Send + Sync + 'static,
+{
+    fn is_block_finalized(&self, raw_hash: H256) -> RpcResult<bool> {
+        let client = self.client.clone();
+        is_block_finalized_inner::<B, C>(self.backend.as_ref(), &client, raw_hash)
+    }
+}
+
+fn is_block_finalized_inner<B: Block<Hash = H256>, C: HeaderBackend<B> + 'static>(
+    backend: &(dyn fc_api::Backend<B>),
+    client: &C,
+    raw_hash: H256,
+) -> RpcResult<bool> {
+    let substrate_hash =
+        match futures::executor::block_on(frontier_backend_client::load_hash::<B, C>(
```

Comment:
I don't understand why are we using an `async_trait` and `block_on` at the same time, but I see moonbeam does the same so I guess just leave it like this...

Reply:
I was kinda wondering the same thing. It seems to work fine without the `async_trait` decorator. Maybe it's ok to remove it?

Reply:
Actually the proper solution is to turn the `fn` into `async fn`, and `block_on` into `await`, but maybe that doesn't work for some reason

Reply:
Updated the code. Pls take a look

Reply:
Perfect, you can propose a similar change to moonbeam

Reply:
https://github.com/moonbeam-foundation/moonbeam/pull/2784